product_supplierinfo_for_customer_sale/models/product_product.py

```diff
@@ -13,14 +13,15 @@ class ProductProduct(models.Model):
     def name_search(self, name='', args=None, operator='ilike', limit=80):
         res = super(ProductProduct, self).name_search(
             name, args=args, operator=operator, limit=limit)
-        if res or not self._context.get('partner_id'):
+        if not name or not self._context.get('partner_id'):
             return res
         partner_id = self._context['partner_id']
         supplierinfo = self.env['product.supplierinfo'].search(
             [('name', '=', partner_id), '|', ('product_code', operator, name),
              ('product_name', operator, name)])
         product = self.search(
-            [('product_tmpl_id', '=', supplierinfo.product_tmpl_id.id)],
+            [('product_tmpl_id', 'in',
+              supplierinfo.mapped('product_tmpl_id').ids)],
             limit=limit)
-        res = product.name_get()
-        return res
+        res.extend(product.name_get())
+        return list(set(res))
```

Comment:
Or better...
Use a `product_ids = set(supplierinfo.mapped('product_tmpl_id').ids) - set([id for id, name in res])` from search call method.

And reduce the limit from current res.
`limit=limit - len(res)`

And you will have unique products with exactly the number of records defined and you will save a lot of time reducing the query.